src/transaction/transaction.rs

```diff
@@ -744,27 +744,60 @@ impl<PdC: PdClient> Transaction<PdC> {
             primary_lock,
             self.timestamp.clone(),
             MAX_TTL,
-            for_update_ts,
+            for_update_ts.clone(),
             need_value,
         );
         let plan = PlanBuilder::new(self.rpc.clone(), request)
             .resolve_lock(self.options.retry_options.lock_backoff.clone())
             .preserve_shard()
-            .retry_multi_region(self.options.retry_options.region_backoff.clone())
+            .retry_multi_region_preserve_results(self.options.retry_options.region_backoff.clone())
             .merge(CollectWithShard)
             .plan();
         let pairs = plan.execute().await;
 
-        // primary key will be set here if needed
-        self.buffer.primary_key_or(&first_key);
+        if let Err(err) = pairs {
+            match err {
+                Error::PessimisticLockError {
+                    inner,
+                    success_keys,
+                } if !success_keys.is_empty() => {
+                    let keys = success_keys.into_iter().map(Key::from);
+                    self.pessimistic_lock_rollback(keys, self.timestamp.clone(), for_update_ts)
+                        .await?;
+                    Err(*inner)
+                }
+                _ => Err(err),
+            }
+        } else {
+            // primary key will be set here if needed
+            self.buffer.primary_key_or(&first_key);
+
+            self.start_auto_heartbeat().await;
 
-        self.start_auto_heartbeat().await;
+            for key in keys {
+                self.buffer.lock(key.key());
+            }
 
-        for key in keys {
-            self.buffer.lock(key.key());
+            pairs
         }
+    }
 
-        pairs
+    /// Rollback pessimistic lock
+    async fn pessimistic_lock_rollback(
```

Comment:
Had better print some logs here.

Reply:
I think we need to appropriately change the buffer state. In the current PR it doesn't affect at all, but I'm afriad the function might be misused. Or we can add a comment to explicitly warn such behavior.

Reply:
> Had better print some logs here.

En... I print a debug msg "rollback pessimistic lock" at the beginning.
What more logs do we need ? *(since there is also a simple debug log at the beginning of other methods)*

Reply:
@ekexium Do we allow the user to retry pessimistic lock? If we do, we should not eagerly rollback the successfully locked keys.

Reply:
> @ekexium Do we allow the user to retry pessimistic lock? If we do, we should not eagerly rollback the successfully locked keys.

Users can retry, but we cannot make assumptions on that. The lock request may already have retried, subject to the backoff option.

Reply:
> 



> > Had better print some logs here.
> 
> En... I print a debug msg "rollback pessimistic lock" at the beginning. What more logs do we need ? _(since there is also a simple debug log at the beginning of other methods)_

I didn't see it 🤦 . The logging is a mess for now. Let's just keep it as is in this PR.

Reply:
> Users can retry, but we cannot make assumptions on that. The lock request may already have retried, subject to the backoff option.

Let me explain my point again. If we allow the user to retry the pessimistic lock when the client returns errors, the client should remember which keys are locked, but not eagerly roll back them. Besides, the client should roll back the locked key when `client.rollback()` is called, or panic if get dropped without rolling back or committing.

Reply:
I think the major drawback of your idea is that if the user doesn't retry the same set of keys, then the locks will remain and may block other txns for some time.

If users want to retry the same set of keys, they can use a different backoff strategy (and we may consider enhance the retry of pessimistic locks to use a newer for_update_ts).

Reference: client-go immediately rollback the successful pessimistic locks, but I think part of the reason is to cooperate with TiDB, client-rust doesn't have to bahave in the same way.

Reply:
I agree with @ekexium . I also think that it would be better to provide an "atomic" method to developers, which will do nothing when return error.